larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -404,6 +425,17 @@ NeutrinoIdTool<T>::SliceFeatures::SliceFeatures(const PfoList &nuPfos, const Pfo
         m_featureVector.push_back(crFracHitsInLongestTrack);
         m_featureVector.push_back(nCRHitsMax);
 
+        m_featureMap["NuNFinalStatePfos"] = nuNFinalStatePfos;
+        m_featureMap["NuNHitsTotal"] = nuNHitsTotal;
+        m_featureMap["NuVertexY"] = nuVertexY;
+        m_featureMap["NuWeightedDirZ"] = nuWeightedDirZ;
+        m_featureMap["NuNSpacePointsInSphere"] = nuNSpacePointsInSphere;
+        m_featureMap["NuEigenRatioInSphere"] = nuEigenRatioInSphere;
+        m_featureMap["CRLongestTrackDirY"] = crLongestTrackDirY;
+        m_featureMap["CRLongestTrackDeflection"] = crLongestTrackDeflection;
+        m_featureMap["CRFracHitsInLongestTrack"] = crFracHitsInLongestTrack;
+        m_featureMap["CRNHitsMax"] = nCRHitsMax;
+
```

Comment:
This is not necessarily a comment for now but it would be 'nice' for this to be automated in such a way that we don't have to manually append lines here if we come up with new features.

Reply:
I agree, I discussed this offline with @AndyChappell but without significant changes to the MvaFeatureVector there didn't seem to be an obvious clean way of achieving this.

Reply:
I guess one way would be not creating the interim variables (like nuFinalStatePfos) and instead storing the calculations straight onto the map.  There would need to be some cleanup functionality to nuke the map if an exception gets thrown though...

Reply:
Yes, in principle we could populate the map as each variable is defined/updated and then populate the feature vector from the map. You do then lose some of the intuitiveness of the variable types from declaring everything as a double, but maybe reduces the chance that you miss a variable in any update?

Reply:
I'm happy to implement this change if you think this would be better. I think in some ways it will becomes less obvious to follow the logic but as you say it may protect from future mistakes in adding variables.

Reply:
I have mixed views about it, would be curious to here @johnmarshall80 's thoughts?

Reply:
Hmmm, can't think of anything great immediately. Could (although it could be easily subverted) just pass the addition to the vector and map over to a member function? E.g. This would take the string and the double, and just do the vector push back and map insertion, via operator[]?

At the moment it doesn't look like any of this has any explicit error handling, but you could of course have the calls to this member function (to add to map and vector) within a try/catch block, with both map and vector wiped clean if there's a relevant exception raised for any one feature?

(Note that BeamParticleId tool currently very similar implementation, but presumably up to e..g ProtoDUNE if they'd like these features persisted too, then would need to follow approach chosen here.)

Reply:
I think this is probably something for a tidying up PR.  I noticed that there's a hardcoded magic number a little about here (line 395 in Henry's PR).  So, there could for sure be a bit of sprucing up.  Addressing this could be done as part of that sprucing.